content/kb/troubleshooting-fstrim-error.md

```diff
@@ -0,0 +1,71 @@
+---
+title: "Troubleshooting: fstrim doesn't work on old kernel"
+author: Phan Le
+draft: false
+date: 2023-10-06
+catelogies:
+  - "trim"
+---
+
+## Applicable versions
+
+* Longhorn version >= `v1.4.0`
+* Node kernel version <= `v4.12`
+
+## Symptoms
+
+When running filesystem trim (either by Longhorn UI or manually by `fstrim` command on the host), it hits the error similar to:
+
+```
+unable to trim filesystem for volume pvc-e381424a-4866-447a-a75c-3096036f7846: cannot find volume pvc-e381424a-4866-447a-a75c-3096036f7846 mount info on host: failed to execute: nsenter [--mount=/host/proc/20357/ns/mnt --net=/host/proc/20357/ns/net fstrim /var/lib/kubelet/plugins/kubernetes.io/csi/driver.longhorn.io/d2df1133f3440486ddec39370380eeed7a3c71499981d63fc80e43c7ca9f4c9e/globalmount], output , stderr fstrim: /var/lib/kubelet/plugins/kubernetes.io/csi/driver.longhorn.io/d2df1133f3440486ddec39370380eeed7a3c71499981d63fc80e43c7ca9f4c9e/globalmount: FITRIM ioctl failed: Input/output error\n: exit status 255
+```
+or
+```
+unable to trim filesystem for volume pvc-e381424a-4866-447a-a75c-3096036f7846: cannot find volume pvc-e381424a-4866-447a-a75c-3096036f7846 mount info on host: failed to execute: nsenter [--mount=/host/proc/20357/ns/mnt --net=/host/proc/20357/ns/net fstrim /var/lib/kubelet/plugins/kubernetes.io/csi/driver.longhorn.io/d2df1133f3440486ddec39370380eeed7a3c71499981d63fc80e43c7ca9f4c9e/globalmount], output , stderr fstrim: /var/lib/kubelet/plugins/kubernetes.io/csi/driver.longhorn.io/d2df1133f3440486ddec39370380eeed7a3c71499981d63fc80e43c7ca9f4c9e/globalmount: the discard operation is not supported : exit status 1
+```
+
+## Details
+
+This is caused by a problem in old kernel SCSI driver which sets the wrong provisioning mode for the SCSI device that supports
+both UNMAP and WRITE SAME capability as detailed in this kernel patch https://github.com/torvalds/linux/commit/bcd069bb250acf6088b60d189ab3ec3ae8dd11a5
+
+### Troubleshooting steps
+
+1. Double-check the engine image of the volume is >= `v1.4.0` in the volume detail page in Longhorn UI. Only engine
+image >= `v1.4.0` supports filesystem trim feature.
+1. If the volume's engine's image was recently live upgraded from a version < `v1.4.0` to >= `v1.4.0`, make sure that the
+volume is detached/reattached first to activate the trim feature in Longhorn volume. You can do this by manually scale
```

Comment:
NIT: Do we need to separate this fstrim failure into 2 cases:
1. Engines are using old IM pod/engine image
2. The node kernel version

Reply:
Quick question, how the instance manager pod can affect trim?

Reply:
For example, an old volume was running in a v1.3.x instance manager pod then got upgraded to engine image v1.4.x. In this case, the upgraded engine was still in the old IM pod, of which the tgt does not support `unmap`.

Reply:
I see. Good point. I separated into 2 cases